### PR TITLE
DOC: stats: Convert sampling tutorial to MyST-md

### PR DESCRIPTION
**Pull request by melissawm**
#### What does this implement/fix?
This PR converts the [Universal Non-Uniform Random Number Sampling in SciPy](https://scipy.github.io/devdocs/tutorial/stats/sampling.html) tutorial to be in [MyST Markdown](https://mystmd.org/) format instead of rst. This is possible by using [MyST-nb](https://myst-nb.readthedocs.io/en/latest/authoring/basics.html#myst-markdown)

#### Additional information
~~Unfortunately, I haven't figured out how to display a "download this file as a notebook" button, but I will keep investigating.~~

EDIT: Deployed version here: https://melissawm.github.io/scipy/tutorial/stats/sampling.html

cc @mdhaber

### CODE REVIEW

**Comment by melissawm** [comment]
See https://github.com/scipy/scipy/issues/20296 for the docs build failure.

**Comment by melissawm** [comment]
Rendered docs link in CircleCI: https://output.circle-artifacts.com/output/job/5defee78-e493-487b-826d-be2ad0c948db/artifacts/0/html/tutorial/stats/sampling.html

**Comment by mdhaber** [comment]
Looks good! We discussed that we'd merge this after investigating:
- use of jupyterlite for execution in browser
- offering a "download as ipynb" link
- use of $ to begin/end math
- removal of >>> from executable code

**Comment by rgommers** [comment]
> use of jupyterlite for execution in browser

@agriyakhetarpal maybe you can help @melissawm with this? Not sure if it's as simple as adding a single directive somewhere, or more would be needed when using MyST + notebooks instead of `.rst` content.

**Comment by agriyakhetarpal** [comment]
@rgommers sure! I think the Notebook directive can be made use of here, I'll pull this branch and investigate how it can be done.

I have previously provided the "Download notebook" button on other prpjects by adding a `_blank` attribute to the HTML `<a>` tag and linking the raw GitHub URL for the notebook.

Another option could to provide an "Open in Google Colab" button if that makes sense? I'm not sure if it works with Markdown notebooks however... see https://githubtocolab.com/ – it is a free service though, and it is possible that there could be some downtime involved someday...

**Comment by melissawm** [comment]
@agriyakhetarpal I was wondering how it's done in [JupyterBook](https://jupyterbook.org/en/stable/intro.html) because that seems like exactly the feature we want (the download button). It probably involves having an extra function on conf.py so that would be my next step. If you can investigate the jupyterlite option, that would be great. Thanks!

**Comment by agriyakhetarpal** [comment]
I have tried adding JupyterLite-based directives to this page – they display for me for a split second and then vanish for no reason, and I could not find anything in `conf.py` that's stopping them. After a bit of time I debugged the HTML source, and I'm finding that each of the buttons:

```html
<button class="try_examples_button hidden"
    onclick="window.tryExamplesShowIframe('e69f12e2-f320-4864-8d51-a6a3ef1d3d3d',
'40491ddc-ba03-42be-9f68-96d280dd6c8b',
'95a4c645-5503-48b5-9cd0-94bc203fb7e7',
'../../lite/tree/../notebooks/?path=945ff25d_7d47_4f74_a48c_081e40be6d09.ipynb','None')">
TEXT TO DEBUG
</button>
```

has a `hidden` attribute (and the `<div>` outside the button that contains it has this too). @steppi, could you perhaps pitch in? This could supposedly be a conflict with any of the other extensions, but TBH that's unlikely.

However, outside of that – if I manually remove the `hidden` text, the button shows up as expected (but without a kernel because `jupyterlite-pyodide-kernel` isn't present as a dependency).

**Comment by steppi** [comment]
> has a `hidden` attribute (and the `<div>` outside the button that contains it has this too). @steppi, could you perhaps pitch in? This could supposedly be a conflict with any of the other extensions, but TBH that's unlikely.

I think it might be because interactive examples are disabled on SciPy right now. It's concerning that it shows up for a split second though, it shouldn't show up at all. See [here](https://jupyterlite-sphinx.readthedocs.io/en/latest/directives/try_examples.html#ignore-patterns). There's a `.try_examples.json` where you can specify ignore patterns. Right now everything is ignored. The `.try_examples.json` in the source directory is a default, but you can modify the `.try_examples.json` in the build directory without rebuilding the docs. You can just change `ignore_patterns` to match everything but the page of interest.

**Comment by agriyakhetarpal** [comment]
Oh you're right, thanks, the `"ignore_patterns": [".*"]` pair seems to be the case. It shows up when you reload the relevant page on a local server I have set up to view the docs, but ~I don't see it on the CircleCI artifacts/previews (maybe it is a latency issue?)~ I do see it there as well but it's quite brief, please refer to [this example for basis spline interpolation](https://output.circle-artifacts.com/output/job/5defee78-e493-487b-826d-be2ad0c948db/artifacts/0/html/reference/generated/scipy.interpolate.BSpline.html#scipy.interpolate.BSpline) and reload the page once or twice?

Modifying the `try_examples.json` in the `build/` directory seems to work!

**Comment by steppi** [comment]
Thanks @agriyakhetarpal, I see that it does appear for a short time. I guess it might make more sense to hide the button initially, and then only remove the hidden class if the page doesn't match ignore patterns.

**Comment by agriyakhetarpal** [comment]
> - use of jupyterlite for execution in browser
> - removal of >>> from executable code

<hr>

Okay, I think I've managed to make things work for the most part, for the first part. The directive was assuming that the Python code snippet was inside a code fence inside a Markdown block, which caused the Pyodide kernel to not load it correctly – but I managed to fix that:

Code snippet containing SciPy's TransformedDensityRejection example

and it embeds the code correctly (I tried to add the diff for the whitespace changes I made, but it is broken here because GFM interprets `\```\` backticks incorrectly). However, outside the notebook, the code snippet is incorrectly added to quote text which I'm investigating further:

Broken code snippet for TransformedDensityRejection example, incorrectly added to quote text

<hr>

As for the second part, i.e., removing `>>>` symbols from executable code – I think we'll have to keep them, otherwise MyST-NB does not interpret code snippets as notebook cells that can be executed, and there may or may not be a workaround for this (I'm yet to find it).

**Comment by melissawm** [comment]
If you find a solution, even if partial, feel free to create a PR to my branch and I'll merge it here. Also happy to incorporate any other suggestions. Thanks!

**Comment by steppi** [comment]
@agriyakhetarpal, if we have notebooks already, then we wouldn't necessarily need to generate the notebook with `try_examples`. `jupyterlite-sphinx` could be updated for `try_examples` to allow a path to a notebook, and then instead of generating one from RST, the button could just swap out whatever is currently rendered under the directive with the embedded notebook.

**Comment by agriyakhetarpal** [comment]
> If you find a solution, even if partial, feel free to create a PR to my branch and I'll merge it here. Also happy to incorporate any other suggestions. Thanks!

@melissawm, I tried some more things and I now have a complete working solution available – it took me a while to find out, but I guess what is happening is `jupyterlite-sphinx` is not parsing code blocks from MyST very well, so I'm now using the [`{eval-rst}`](https://myst-parser.readthedocs.io/en/latest/syntax/roles-and-directives.html#how-directives-parse-content) directive to parse the code block within reStructuredText, and then pass it to `.. try_examples::` which properly receives code from code blocks. I'll send a PR your way to target this branch, along with a few ancillary changes. This is essentially a workaround because the example isn't running in MySt, but seems to work very nicely.

> if we have notebooks already, then we wouldn't necessarily need to generate the notebook with `try_examples`. `jupyterlite-sphinx` could be updated for `try_examples` to allow a path to a notebook, and then instead of generating one from RST, the button could just swap out whatever is currently rendered under the directive with the embedded notebook.

@steppi, I'm not sure I completely understand what you meant by that, actually – do you mean that the `try_examples` can be made to attempt to add itself to all legible examples in a notebook file, say, through the top of the file? If yes, I think that should be a nice feature to add.

**Comment by steppi** [comment]
> @steppi, I'm not sure I completely understand what you meant by that, actually – do you mean that the `try_examples` can be made to attempt to add itself to all legible examples in a notebook file, say, through the top of the file? If yes, I think that should be a nice feature to add.

Sorry; nevermind. I wasn't paying attention. I thought there was a notebook file that was being parsed and turned into RST with myst-md. That's not what's going on here though.

**Comment by melissawm** [comment]
Ok - so from @agriyakhetarpal 's PR to my PR (linked above) it looks like it's an either/or situation: either we have an executed notebook with all outputs embedded, or we have the "try with jupyterlite" button. It makes sense actually - either you execute the code on build time, or the user does it when they are visiting the page. 

I don't know what would be the ideal solution here right now. For the future, we could think of a hybrid option that would provide executed notebooks and still have the try with jupyterlite button, but I'd like to propose decoupling that from this PR and investigating further. 

What do you think @mdhaber ?

**Comment by agriyakhetarpal** [comment]
Thanks to the discusion that entailed, we are probably one step closer to getting the notebook running in a tab, and also towards providing a method to downloading the notebook as mentioned in https://github.com/scipy/scipy/pull/20303#issuecomment-2016531823, given that it will be stored as a static asset after the documentation has been generated. It should be a matter of extracting the link and making it available with a button – and the link can then be added in an admonition. I have previously used `nbsphinx`'s prolog configuration value for this and neatly inserted the admonition as raw HTML ([example](https://docs.pybamm.org/en/stable/source/examples/notebooks/models/MPM.html)). Therefore, a similar method could be to add a template HTML file that can be applied to an appropriate location on the page. I know that PST allows this as mentioned in https://pydata-sphinx-theme.readthedocs.io/en/stable/user_guide/layout.html#article-footer, so we'll have to enable this conditionally (maybe add the template to the `html_theme_options` dictionary only if it is a `scipy.stats.*` page or so – but I'm sure there can be better ways to handle this).

**Comment by melissawm** [comment]
Hello folks - I think this is ready for another round. In discussing with @agriyakhetarpal we could not come with a better solution for now than explicitly converting the md file to an ipynb file before building the docs. This works, but I'm not very happy with this solution since I would prefer to not touch the Makefile and/or not have duplicate work done.

In this current iteration, we generate a local ipynb file right next to sampling.md, and when the docs are built we generate another sampling.ipynb file, which is used by myst-nb to generate the html pages. If there are any suggestions for a better workflow, I'd be happy to hear.

For some reason, the jupyterlab session does not come up on CI (I assume it's been turned off). So here's how this page looks like in my fork, with the "Open as a notebook" button fully functional: https://melissawm.github.io/scipy/tutorial/stats/sampling.html

cc @mdhaber

**Comment by agriyakhetarpal** [comment]
Thanks for tagging me, @melissawm – from a hasty look, this looks much better than my (similar and non-Makefile) solution in PyWavelets/pywt#741 – the notebook loads quite neatly in a new tab! It's missing a kernel, though. I have had problems with removing the directives that were inserted at the top of the notebook, so I guess they'll have to stay for now.

**Comment by mdhaber** [comment]
> In this current iteration, we generate a local ipynb file right next to sampling.md, and when the docs are built we generate another sampling.ipynb file, which is used by myst-nb to generate the html pages. 

IIUC, in both cases the `ipynb` file is temporary? It is generated from the `md` as an intermediate step, whether the desired output is an HTML file or for viewing the tutorial as a notebook?

**Comment by melissawm** [comment]
Yes @mdhaber  - the ipynb files will be there for the deployed website, but won't live in the repo. I added them to the .gitignore file.

Marking as ready for review then 😄 Thanks!

**Comment by mdhaber** [comment]
I'm not sure if doc build time is as important as other build times, but in case it is important to others, it looks like doc build time is no longer in this PR than main (based on [5b2dae4](https://github.com/scipy/scipy/pull/20303/commits/5b2dae4f6260816838c700cb1a29cf6c04ce2b69)).

File: doc/source/dev/contributor/rendering_documentation.rst [hunk 1]
```diff
@@ -102,11 +102,11 @@ on the cloud.
 Adding or editing tutorials as Jupyter notebooks
 ------------------------------------------------
 
-Under the ``doc/source/notebooks/`` folder of the SciPy tree you can find a few
+Under the ``doc/source/`` folder of the SciPy tree you can find a few
```
Comment (mdhaber) [inline doc/source/dev/contributor/rendering_documentation.rst:105]: Does this mean that they can be anywhere within the `doc/source` folder now? Is there anything to add here, or will an `md` file added in this folder "just work"? Where is the Interpolate transition guide now?

Comment (melissawm) [reply doc/source/dev/contributor/rendering_documentation.rst:105]: The .md files will "just work" (be converted as regular html) no matter their location in the doc/source tree. For the "Open as notebook" button to work, .md files will have to be converted to .ipynb files first. Currently, only the ones in the stats folders will be converted but we can add more later. The interpolate transition guide was transformed into a static .md file in #20497

**Review by mdhaber — APPROVED** [review]
I can't review how it's working, just how it can be edited and how appears, and I'm happy with that. 

The rendered documentation is very similar, the most obvious difference being the appearance of the "Open as a notebook button".

I like the colors of the existing code rendering a bit better,
image

but I like how this does not have the `>>>` and `...` - especially in the source markdown.
image

Copy-pasting these symbols into my IDE does not work properly, and generating them in the first place is annoying to have to do (and the cause of some draft PR bugs, where `>>>` is used in place of `...`), so this will save some time and frustration. (Of course, the rest of the markdown syntax is much better than ReST, too; that goes without saying. Thanks for getting `$` to render math!)

Another difference is that there is physical separation between code input and code output:

Here:
image

vs main:
image

It would be nice if this could be a little more compact - ideally like it appears in the notebook:
image

but I can definitely live with it.

The plot is bigger now, too, although I suppose that is something that could always have been adjusted with settings.

For the notebook itself, it looks like the stuff at the top is unavoidable for now:
image

Admonitions don't render well:
image

Nor do links and other references:
image

But I imagine these can improve in time. It's much better to have the notebooks than to exclude them because they're not yet pristine.

File: doc/source/dev/contributor/rendering_documentation.rst (hunk 1 above)
Comment (mdhaber) [reply doc/source/dev/contributor/rendering_documentation.rst:105]: I see:
image

OK, thanks!

**Comment by mdhaber** [comment]
> For some reason, the jupyterlab session does not come up on CI (I assume it's been turned off).

I remember that executable docstring examples were enabled for a bit, then turned off. Does it have something to do with that? Does this mean that when this PR is merged, we still won't see the button, or do you think it's just a CI thing?

**Comment by agriyakhetarpal** [comment]
> > For some reason, the jupyterlab session does not come up on CI (I assume it's been turned off).
> 
> I remember that executable docstring examples were enabled for a bit, then turned off. Does it have something to do with that? Does this mean that when this PR is merged, we still won't see the button, or do you think it's just a CI thing?

Those buttons are [disabled through `jupyterlite-sphinx`'s configuration file](https://github.com/scipy/scipy/blob/09769690f65cfc078c860f960f084aa3689df5b4/doc/source/try_examples.json#L3) for now. I have been working on updating the SciPy recipe to v1.13.0 in-tree for Pyodide (pyodide/pyodide#4719), and after a Pyodide patch release with it, they should be at a better position to be enabled again.

**Review by rgommers — COMMENTED** [review]
> but I think it's the jupyterlite generation that doesn't run on CI because it has been disabled except for local builds.

So the "Open as a notebook" button should be controlled by the same switch that @steppi added for the other interactive docs components.

Locally I have all the tools installed, but I don't see the button. In the CI preview I do see the button, but it doesn't work as you say. So something looks inverted here.

* * *

This seems fast enough. The debugger warning output would ideally not show up, so try setting `PYDEVD_DISABLE_FILE_VALIDATION` like the warning suggests?

```
/home/rgommers/code/scipy/doc/source/tutorial/stats/sampling.md: Executing notebook using local CWD [mystnb]
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
/home/rgommers/code/scipy/doc/source/tutorial/stats/sampling.md: Executed notebook in 1.82 seconds 
```

* * * 

Other than that, things look pretty polished to me. The changes to `environment.yml` and doc config files look fine.

**Comment by melissawm** [comment]
To summarize, the missing bits here are:
- [ ] Figuring out how to 1) turn on jupyterlite for a specific build and 2) Remove the button if jupyterlite is NOT enabled for that build;
- [ ] Silence jupyter debugging warnings

**Comment by melissawm** [comment]
I think this is finally ready 😄

**Review by steppi — APPROVED** [review]
This looks good to me. @melissawm  and I worked out a way to make the warning about having both `sampling.md` and `sampling.ipynb` go away by storing the notebook in the `source/_contents` folder. This folder is also used by jupyterlite-sphinx for storing notebooks and gets cleaned up by `jupyterlite-sphinx` when it is finished. This means we don't have to worry about cleaning notebook files like the one in this PR ourselves.

All in all, I think this is really cool. I'd like to add a feature to jupyterlite-sphinx to use markdown notebooks like this.

**Pull request merged**